### PR TITLE
Feature/ops wcoss2 transition to wcoss2 for bufr sounding component

### PR DESCRIPTION
**Pull request by GuangPingLou-NOAA**
Bufr sounding component of the package is ready for merger. Scripts have been modified to accommodate the requirements of wcoss2 environments, including modules and resources, etc. All tests have been done, including a job card submissions and rocoro run workflow. Results were examined and validated.

### CODE REVIEW

File: jobs/rocoto/postsnd.sh [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
+#!/bin/sh -x
 
 ###############################################################
 # Source FV3GFS workflow modules
 
-. $HOMEgfs/ush/load_fv3gfs_modules.sh
+##. $HOMEgfs/ush/load_fv3gfs_modules.sh
+. $HOMEgfs/versions/run.ver
+
+. $HOMEgfs/modulefiles/module_base.wcoss2
```
Comment (KateFriedman-NOAA) [inline jobs/rocoto/postsnd.sh]: Change this back to the load_fv3gfs_module.sh line. I have updates coming in with my PR into feature/ops-wcoss2 that update and correct the module_base loading at runtime from load_fv3gfs_module.sh. I'm hoping to submit that PR first and then you can update your fork branch with those updates.

File: sorc/build_gfs_bufrsnd.sh [hunk 2]
```diff
@@ -5,6 +5,7 @@ source ./machine-setup.sh > /dev/null 2>&1
 cwd=`pwd`
 
 USE_PREINST_LIBS=${USE_PREINST_LIBS:-"true"}
+source ../versions/build.ver
```
Comment (KateFriedman-NOAA) [inline sorc/build_gfs_bufrsnd.sh]: Source machine-setup.sh here instead, the machine-setup.sh script source build.ver for you. See other build scripts for examples.

File: jobs/rocoto/postsnd.sh (hunk 1 above)
Comment (GuangPingLou-NOAA) [reply jobs/rocoto/postsnd.sh]: Went back to loading the original  load_fv3gfs_module.sh. I'll wait for your commit and then pull it  to my branch.

File: sorc/build_gfs_bufrsnd.sh (hunk 2 above)
Comment (GuangPingLou-NOAA) [reply sorc/build_gfs_bufrsnd.sh]: Changed it. Thanks.

**Comment by KateFriedman-NOAA** [comment]
@GuangPingLou-NOAA Please sync your fork branch now that I have submitted PR #490. Let us know when you're incorporated the updates and resolved any conflicts. I'll do the review when you're ready, thanks!

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA Do I need to open a pull request so that I can fetch the upstream (authoritative branch) and then resolve conflicts?

**Comment by KateFriedman-NOAA** [comment]
@GuangPingLou-NOAA Also, please convert `modulefiles/gfs_bufr.wcoss2` into a LUA modulefile. 

See the conversion I just did on module_base.wcoss2:
https://github.com/NOAA-EMC/global-workflow/blob/feature/ops-wcoss2/modulefiles/module_base.wcoss2.lua

Let me know if you have any questions about that conversion, thanks!

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA Do I need to open a pull request so that I can fetch the upstream (authoritative branch) and then resolve conflicts?

You can do a sync merge two ways:

1. use the "Fetch and merge" button in GitHub when looking at your fork branch (if no conflicts exist)
2. do the sync merge manually on the command line (if conflicts exist)

For option 2 see this part of the g-w wiki for commands to use: https://github.com/NOAA-EMC/global-workflow/wiki/Development-of-Global-Workflow#how-to-sync-fork-with-the-authoritative-repository

Let me know if you have questions about those steps or run into issues. Feel free to ping me in Slack for sync questions. Thanks!

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA I think the conflicts have been resolved.

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA In the script "load_fv3gfs_modules.sh", it remains this line: 
        "module load module_base.wcoss2". Do we need to change it to "module load module_base.wcoss2.lua"?

**Comment by KateFriedman-NOAA** [comment]
@GuangPingLou-NOAA Yep, I don't see a conflict message in the PR anymore.

> @KateFriedman-NOAA In the script "load_fv3gfs_modules.sh", it remains this line: "module load module_base.wcoss2". Do we need to change it to "module load module_base.wcoss2.lua"?

I'm getting to know LUA modules myself still but from what I'm seeing so far you don't need to reference the ".lua" part when loading it (found that by trial and error). @aerorahul @WalterKolczynski-NOAA Can you clarify the need for including the ".lua" bit in the file name? Wanna make sure we understand it moving forward, thanks!

**Review by KateFriedman-NOAA — CHANGES_REQUESTED** [review]
Changes look good, will need the modulefile converted to LUA format though, awaiting clarification from other developers on that.

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA without .lua attached to the filename, module_base.wcoss2, it would not be recognized when executing it. It will say "No such file ...."

**Comment by KateFriedman-NOAA** [comment]
> without .lua attached to the filename, module_base.wcoss2, it would not be recognized when executing it. It will say "No such file ...."

@GuangPingLou-NOAA Let's wait for the others to comment, I was able to run without including the ".lua" bit.

**Comment by GuangPingLou-NOAA** [comment]
Sure. Thanks.

**Comment by KateFriedman-NOAA** [comment]
@GuangPingLou-NOAA I just submitted @BoiVuong-NOAA's PR into feature/ops-wcoss2. Please update your fork branch again. Your PR is next in line to go into feature/ops-wcoss2 after we get clarification on LUA stuff.

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA Updated my fork already. Thanks.

File: ush/load_fv3gfs_modules.sh [hunk 3]
```diff
@@ -34,7 +34,7 @@ elif [[ -L /usrx && "$( readlink /usrx 2> /dev/null )" =~ dell ]] ; then
 	module load module_base.wcoss_dell_p3 
 elif [[ -d /lfs/h2 ]]; then
     # We are on WCOSS2 (Cactus or Dogwood)
-	module load module_base.wcoss2
+	module load module_base.wcoss2.lua
```
Comment (KateFriedman-NOAA) [inline ush/load_fv3gfs_modules.sh]: @GuangPingLou-NOAA Please remove the `.lua` part on line 37 and I will submit this PR. I'll convert the module file to LUA when I convert the other modulefiles and then resolve any issues related to that. Thanks!

**Review by KateFriedman-NOAA — CHANGES_REQUESTED** [review]
Remove the `.lua` text on line 37 in load_fv3gfs_modules.sh, thanks!

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA Just got around to update "module_base.wcoss2.sh". Thanks!

**Review by BoiVuong-NOAA — COMMENTED** [review]
Just a question
GuangPing,
Will you convert all modulefiles *.wcoss2 files  to .lua format ?
Thanks,
Boi

**Comment by GuangPingLou-NOAA** [comment]
@BoiVuong-NOAA No, Boi. That's the only place I change it to .lua. Kate knows more about converting module load to .lua.

**Comment by BoiVuong-NOAA** [comment]
I see. Thanks, GuangPing

**Comment by KateFriedman-NOAA** [comment]
@BoiVuong-NOAA I am going to convert the remaining wcoss2 module files next.

@GuangPingLou-NOAA Thanks! Will submit shortly.

**Review by KateFriedman-NOAA — APPROVED** [review]
Looks good, thanks!

**Comment by KateFriedman-NOAA** [comment]
@BoiVuong-NOAA @GuangPingLou-NOAA Please sync your fork branches, several PRs have gone in and another is on the way. Thanks!

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA Great, Kate! Thank you so much! Please let me know if you encounter any issues running the component,

**Comment by BoiVuong-NOAA** [comment]
Yes, I will.  Do you have a chance to modify machine-setup.sh file  ?

**Comment by KateFriedman-NOAA** [comment]
> Please let me know if you encounter any issues running the component,

@GuangPingLou-NOAA Will do, hope to test the job in the next week. Will have you look at the log and output when I get it run.

> Do you have a chance to modify machine-setup.sh file ?

@BoiVuong-NOAA Not yet, sorry. Was on leave all last week. Will try to resolve that issue this week.

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA machine-setup.sh added this portion:
elif [[ -d /lfs/h2 ]] ; then
    # We are on NOAA Cactus or Dogwood
    if ( ! eval module help > /dev/null 2>&1 ) ; then
        echo load the module command 1>&2
        source /usr/share/lmod/lmod/init/$__ms_shell
    fi
    target=wcoss2
    module purge
    module load envvar/1.0

It works for me,

**Comment by BoiVuong-NOAA** [comment]
Thanks, Kate.   I will modify the  machine-setup.sh file.

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA machine-setup.sh added this portion: elif [[ -d /lfs/h2 ]] ; then # We are on NOAA Cactus or Dogwood if ( ! eval module help > /dev/null 2>&1 ) ; then echo load the module command 1>&2 source /usr/share/lmod/lmod/init/$__ms_shell fi target=wcoss2 module purge module load envvar/1.0
> 
> It works for me,

@GuangPingLou-NOAA The issue @BoiVuong-NOAA is referring to is related to the gfs util build script getting a path mismatch when it runs because it's at a different level than the machine-setup.sh script (related to sourcing the new build.ver file). I have to figure out how to introduce the correct path without breaking machine-setup.sh for the other builds and at runtime. The machine-setup.sh script works otherwise, as you noted.

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA Alright, understood. Thanks!

**Pull request merged**